client/peerset/src/protocol_controller.rs

```diff
@@ -0,0 +1,576 @@
+	reserved_only: bool,
+	/// Peers and their connection states (including reserved nodes).
+	nodes: HashMap<PeerId, ConnectionState>,
+	/// Next time to allocate slots. This is done once per second.
+	next_periodic_alloc_slots: Instant,
+	/// Outgoing channel for messages to `Notifications`.
+	to_notifications: TracingUnboundedSender<Message>,
+}
+
+impl ProtocolController {
+	/// Construct new [`ProtocolController`].
+	pub fn new(
+		set_id: SetId,
+		config: SetConfig,
+		to_notifications: TracingUnboundedSender<Message>,
+	) -> (ProtocolHandle, ProtocolController) {
+		let (to_controller, from_handle) = tracing_unbounded("mpsc_protocol_controller", 10_000);
+		let handle = ProtocolHandle { to_controller };
+		let nodes = config
+			.reserved_nodes
+			.union(&config.bootnodes.into_iter().collect())
+			.map(|p| (*p, ConnectionState::NotConnected))
+			.collect::<HashMap<PeerId, ConnectionState>>();
+		let controller = ProtocolController {
+			set_id,
+			from_handle,
+			num_in: 0,
+			num_out: 0,
+			max_in: config.in_peers,
+			max_out: config.out_peers,
+			reserved_nodes: config.reserved_nodes,
+			reserved_only: config.reserved_only,
+			nodes,
+			next_periodic_alloc_slots: Instant::now(),
+			to_notifications,
+		};
+		(handle, controller)
+	}
+
+	/// Drive [`ProtocolController`]. This function returns when all instances of
+	/// [`ProtocolHandle`] are dropped.
+	pub async fn run(mut self) {
+		while self.next_action().await {}
+	}
+
+	/// Perform one action. Returns `true` if it should be called again.
+	///
+	/// Intended for tests only. Use `run` for driving [`ProtocolController`].
+	pub async fn next_action(&mut self) -> bool {
+		let action = loop {
+			let mut next_alloc_slots = Delay::new_at(self.next_periodic_alloc_slots).fuse();
+			futures::select! {
+				action = self.from_handle.next() => match action {
+					Some(action) => break action,
+					None => return false,
+				},
+				_ = next_alloc_slots => {
+					self.alloc_slots();
+					self.next_periodic_alloc_slots = Instant::now() + Duration::new(1, 0);
+				},
+			}
+		};
+		match action {
+			Action::AddReservedPeer(peer_id) => self.on_add_reserved_peer(peer_id),
+			Action::RemoveReservedPeer(peer_id) => self.on_remove_reserved_peer(peer_id),
+			Action::SetReservedPeers(peer_ids) => self.on_set_reserved_peers(peer_ids),
+			Action::SetReservedOnly(reserved_only) => self.on_set_reserved_only(reserved_only),
+			Action::AddPeer(peer_id) => self.on_add_peer(peer_id),
+			Action::RemovePeer(peer_id) => self.on_remove_peer(peer_id),
+			Action::IncomingConnection(peer_id, index) =>
+				self.on_incoming_connection(peer_id, index),
+		}
+		true
+	}
+
+	fn on_add_reserved_peer(&mut self, peer_id: PeerId) {
+		if !self.reserved_nodes.insert(peer_id) {
+			return
+		}
+		// Discount occupied slots or connect to the node.
+		match self.nodes.entry(peer_id).or_default() {
+			ConnectionState::In => self.num_in -= 1,
+			ConnectionState::Out => self.num_out -= 1,
```

Comment:
Shouldn't it be that reserved peers don't update `num_in`/`num_out` as they don't occupy slots whereas normal connections update these? Now the logic behind updating these counters seem backwards because `on_add_peer()`/`on_remove_peer()` don't update increase/decrease the counters.

Reply:
Once a peer becomes reserved, it stops occupying a regular slot, so we "free" one slot. `on_add_peer()`/`on_remove_peer()` also update the slot counters, but via the `Peer` abstraction. (And I don't like that in some places we use this abstraction, and in some just do things manually.)